### PR TITLE
Add YOLO Detr wrapper support

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

Related to https://github.com/open-edge-platform/geti/issues/7365

### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [ ] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: library/src/getitune/backend/ultralytics/models/base.py [hunk 1]
```diff
@@ -144,6 +140,10 @@ def yolo(self) -> YOLO:
 
     def _build_yolo(self) -> YOLO:
         """Create the ``ultralytics.YOLO`` model and optionally load pretrained weights."""
+        if self.model_name.endswith(".pt"):
+            logger.info(f"Building Ultralytics model from checkpoint: {self.model_name} (task={self.task})")
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/models/base.py:149]: What if `self.pretrained = False`? Previously there was a validation check that raised an error when `self.pretrained = False` when initializing through checkpoint.

File: library/src/getitune/backend/ultralytics/trainers/classification.py [hunk 2]
```diff
@@ -84,7 +84,7 @@ def get_validator(self) -> ClassificationValidator:
         if not self._use_getitune_data:
             return super().get_validator()  # type: ignore[return-value]
 
-        self.loss_names = ["loss"]
+        self.loss_names = ["loss"]  # pyrefly: ignore[bad-assignment]
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/trainers/classification.py]: Why does Pyrefly warn about a `bad-assignment` here? And why is `loss_names` defined in this function rather than being set in the constructor?

File: library/src/getitune/backend/ultralytics/trainers/yolo_detr.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from getitune.backend.ultralytics.validators.yolo_detr import YoloDetrValidator
+
+from .base import GetiTuneBaseTrainer
+
+if TYPE_CHECKING:
+    from torch import nn
+    from ultralytics.models.rtdetr.val import RTDETRValidator as _RTDETRValidator
+
+
+class YoloDetrTrainer(  # pyrefly: ignore[inconsistent-inheritance]
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/trainers/yolo_detr.py]: Why does Pyrefly warn about `inconsistent-inheritance` here?

File: library/src/getitune/backend/ultralytics/models/criterion.py [hunk 4]
```diff
@@ -23,17 +23,17 @@ def forward(
...
 
         Returns:
-            Tuple of (loss, detached loss).
+            Tuple of (loss, dict of detached per-component losses).
         """
         logits = preds[0] if isinstance(preds, (list, tuple)) else preds
         targets = batch["cls"].float()
         loss = functional.binary_cross_entropy_with_logits(logits, targets, reduction="mean")
-        return loss, loss.detach()
+        return loss, {"loss": loss.detach()}
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/models/criterion.py:39]: Is this a bugfix?

File: library/src/getitune/backend/ultralytics/models/semantic_segmentation.py [hunk 5]
```diff
@@ -5,21 +5,68 @@
...
+@contextmanager
+def _force_logits_semantic_export() -> Iterator[None]:
+    """Temporarily disable the Ultralytics semantic head's baked class map.
+
+    Ultralytics branches the ``SemanticSegment`` head on the export format: for
+    ``onnx``/``mnn``/``openvino`` it bakes ``argmax`` into the graph and emits a
+    single-channel ``[B, H, W]`` class map. ModelAPI's ``SegmentationModel``
+    cannot consume that layout — it misclassifies the 3D output as an
+    ``[H, W, num_classes]`` tensor with ``num_classes == 1`` and fails on
+    ``np.argmax(soft_prediction, axis=2)``. getitune's export contract
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/models/semantic_segmentation.py:37]: Is this something that should be maybe addressed in ModelAPI? I feel like ModelAPI wrappers set very strict expectations on the model output shape, adding friction when we (or contributors) want to integrate models from 3rd party frameworks. If ModelAPI classes exposed e.g. a parameter `layout: GraphOutputLayout` with values `GraphOutputLayout.BHW`, `GraphOutputLayout.HWC`, ... then they could internally resolve the differences without requiring the client to implement hacks like this.

File: library/src/getitune/backend/ultralytics/models/base.py (hunk 1 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/models/base.py:149]: I don't see the reason to raise an error when self.pretrained = False and load from .pt
However, I can add a warning, that self.pretrained parameter will be ignored

File: library/src/getitune/backend/ultralytics/models/criterion.py (hunk 4 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/models/criterion.py:39]: Yeah, they changed interface ...

File: library/src/getitune/backend/ultralytics/models/semantic_segmentation.py (hunk 5 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/models/semantic_segmentation.py:37]: Yes, if they won't change it after the official release (a parameter that controls argmax during export), we should add support of this in modelAPI. 
For now, I would leave this work around as is, since we don't know the final release state.

This PR introduces a general support for YOLODetr to enable YOLO27 fast in the same day with official release. All bug fixes and refactoring can be done after that.

File: library/src/getitune/backend/ultralytics/trainers/yolo_detr.py (hunk 3 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/trainers/yolo_detr.py]: Removed

File: library/src/getitune/backend/ultralytics/trainers/detection.py [hunk 6]
```diff
@@ -42,7 +42,7 @@ def get_validator(self) -> _UltralyticsDetectionValidator:
         if not self._use_getitune_data:
             return super().get_validator()
 
-        self.loss_names = ["box_loss", "cls_loss", "dfl_loss"]
+        self.loss_names = ["box_loss", "cls_loss", "dfl_loss"]  # pyrefly: ignore[bad-assignment]
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/trainers/detection.py:45]: Same question as in my previous comment, also for instance/semantic segmentation: why is `loss_names` assigned in `get_validator` instead of the constructor, resulting in a `bad-assignment` typing warning?

File: library/src/getitune/backend/ultralytics/trainers/yolo_detr.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            return _DEIMTrainer.preprocess_batch(self, batch)
+        return self._move_batch_to_device(batch)
+
+    def train(self, *args: Any, **kwargs: Any) -> Any:  # noqa: ANN401
+        """Skip upstream epoch callbacks that require its native dataset type."""
+        if not self._use_getitune_data:
+            return super().train(*args, **kwargs)
+
+        if self.args.close_mosaic:
+            self.args.close_mosaic = 0
```
Comment (leoll2) [inline library/src/getitune/backend/ultralytics/trainers/yolo_detr.py]: Can you explain this line?

File: library/pyproject.toml [hunk 8]
```diff
@@ -187,6 +187,8 @@ torchvision = [
     { index = "torch-xpu", extra = "xpu"},
     { index = "torch-cuda", extra = "cuda"},
 ]
+ultralytics = { git = "https://github.com/ultralytics/ultralytics.git", branch = "exp27-clean" }
```
Comment (leoll2) [inline library/pyproject.toml]: Beware that the `exp27-clean` branch may be deleted as soon as it's merged into `main`. Before merging this PR, perhaps it's worth switching to a fixed commit (-> the most recent from `exp27-clean`) to prevent the possibility that the CI could break overnight.

File: library/pyproject.toml [hunk 9]
```diff
@@ -187,6 +187,8 @@ torchvision = [
...
+openvino-model-api = { git = "https://github.com/open-edge-platform/model_api.git", branch = "master", subdirectory = "model_api" }
```
Comment (leoll2) [inline library/pyproject.toml:191]: Could you also update this line:

https://github.com/open-edge-platform/geti/blob/0a46b1501be022bfa6dc3b5722f03b683ec4e449/application/backend/app/services/demo_files_service.py#L366

to `"openvino-model-api[onnx] @ git+https://github.com/open-edge-platform/model_api@master#subdirectory=model_api"` ?

This will ensure that the inference demo script works even if someone builds Geti from source, trains a YOLO27 model and exports it.

File: library/src/getitune/backend/ultralytics/trainers/yolo_detr.py (hunk 7 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/trainers/yolo_detr.py]: We switch off close mosaic parameter when use getitune data layer, because we control augmentations on our side.
However, I could just set it to 0 without check

File: library/pyproject.toml (hunk 8 above)
Comment (kprokofi) [reply library/pyproject.toml]: Yes, fair point

File: library/src/getitune/backend/ultralytics/trainers/detection.py (hunk 6 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/trainers/detection.py:45]: Two reasons why it can't move to the constructor:

1. The names are not always static. For YOLO-DETR they depend on the model head (DeimDecoder adds fgl_loss/ddf_loss), and the model is only built later, inside _setup_train(). get_validator() runs right after the model is built, so it is the first place where the head is known.

2. The typing warning would stay either way. Upstream BaseTrainer.__init__ does self.loss_names = (), so the type checker infers the attribute as tuple[()], and a subclass cannot narrow a read-write attribute type. Assigning in the constructor fires the exact same warning, so moving it does not remove the pyrefly: ignore.

We also keep the assignment there because the validator's metric keys (label_loss_items) read loss_names immediately after get_validator() returns, so the names must be final at that point.

And yes, in upstream they assign loss names in validator as well :)
Upstream RTDETRTrainer.get_validator assigns self.loss_names inside get_validator from the model head
Ultralytics code is "slightly not ideal" sometimes.

File: library/pyproject.toml (hunk 8 above)
Comment (kprokofi) [reply library/pyproject.toml]: They already delete it...

File: library/src/getitune/backend/ultralytics/trainers/yolo_detr.py (hunk 7 above)
Comment (kprokofi) [reply library/src/getitune/backend/ultralytics/trainers/yolo_detr.py]: Done

**Review by leoll2 — DISMISSED** [review]

**Review by leoll2 — APPROVED** [review]
